File: algorithms/0099.recover-binary-search-tree/solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def recover_tree(root)
  @vals = []
  nodeVals(root)
  @vals = @vals.sort.reverse
  recover(root)
end

def nodeVals(node)
  return if node.nil?
  @vals << node.val
  nodeVals(node.left)
  nodeVals(node.right)
end

def recover(node)
  return if node.nil?

  recover(node.left)
  node.val = @vals.pop
  recover(node.right)
end

File: algorithms/0099.recover-binary-search-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func recoverTree(root *TreeNode) {
	vals := []int{}
	nodeVals(root, &vals)
	sort.Ints(vals)

	fmt.Println(vals[0])
	recover(root, &vals)
}

func nodeVals(node *TreeNode, vals *[]int) {
	if node == nil {
		return
	}

	*vals = append(*vals, node.Val)
	nodeVals(node.Left, vals)
	nodeVals(node.Right, vals)
}

func recover(node *TreeNode, vals *[]int) {
	if node == nil {
		return
	}

	recover(node.Left, vals)

	node.Val = (*vals)[0]
	*vals = (*vals)[1:]

	recover(node.Right, vals)
}
